lib/ex_webrtc/rtp_receiver/report_recorder.ex

```diff
@@ -0,0 +1,201 @@
+defmodule ExWebRTC.RTPReceiver.ReportRecorder do
+  @moduledoc false
+
+  import Bitwise
+
+  alias ExRTCP.Packet.{ReceiverReport, ReceptionReport}
+
+  @max_u32 0xFFFFFFFF
+  @max_u24 0xFFFFFF
+  @max_seq_no 0xFFFF
+  @breakpoint 0x7FFF
+
+  @type t() :: %__MODULE__{
+          sender_ssrc: non_neg_integer(),
+          media_ssrc: non_neg_integer(),
+          clock_rate: non_neg_integer(),
+          lost_packets: MapSet.t(),
+          last_seq_no: {non_neg_integer(), ExRTP.Packet.uint16()},
+          last_report_seq_no: {non_neg_integer(), ExRTP.Packet.uint16()},
+          last_rtp_timestamp: ExRTP.Packet.uint32(),
+          last_timestamp: integer(),
+          last_sr_ntp_timestamp: ExRTP.Packet.uint32(),
+          last_sr_timestamp: integer(),
+          jitter: float(),
+          total_lost: non_neg_integer()
+        }
+
+  @enforce_keys [:sender_ssrc, :media_ssrc, :clock_rate]
+  defstruct [
+              lost_packets: MapSet.new(),
+              last_seq_no: nil,
+              last_report_seq_no: nil,
+              last_rtp_timestamp: nil,
+              last_timestamp: nil,
+              last_sr_ntp_timestamp: 0,
+              last_sr_timestamp: nil,
+              jitter: 0,
+              total_lost: 0
+            ] ++ @enforce_keys
+
+  @doc """
+  Records incoming RTP Packet.
+  `time` parameter accepts output of `System.monotonic_time(:native)` as a value.
+  """
+  @spec record_packet(t(), ExRTP.Packet.t(), integer()) :: t()
+  def record_packet(%{last_seq_no: nil} = recorder, packet, time) do
+    # seq_no == {cycle_no, seq_no as in RTP packet}
+    %__MODULE__{
+      recorder
+      | last_seq_no: {0, packet.sequence_number},
+        last_report_seq_no: {0, packet.sequence_number - 1},
+        last_rtp_timestamp: packet.timestamp,
+        last_timestamp: time
+    }
+  end
+
+  def record_packet(recorder, packet, time) do
+    recorder
+    |> record_seq_no(packet.sequence_number)
+    |> record_jitter(packet.timestamp, time)
+  end
+
+  @doc """
+  Records incoming RTCP Sender Report.
+  `time` parameter accepts output of `System.monotonic_time(:native)` as a value.
+  """
+  @spec record_report(t(), ExRTCP.Packet.SenderReport.t(), integer()) :: t()
+  def record_report(recorder, sender_report, time) do
+    # we take the middle 32 bits of the NTP timestamp
+    ntp_ts = sender_report.ntp_timestamp >>> 16 &&& @max_u32
+
+    %__MODULE__{recorder | last_sr_ntp_timestamp: ntp_ts, last_sr_timestamp: time}
+  end
+
+  @doc """
+  Creates an RTCP Receiver Report.
+  `time` parameter accepts output of `System.monotonic_time(:native)` as a value.
+  """
+  @spec get_report(t(), integer()) :: {ReceiverReport.t(), t()}
+  def get_report(recorder, time) do
+    received =
+      recorder.last_seq_no
+      |> seq_no_diff(recorder.last_report_seq_no)
+      |> min(@max_u24)
```

Comment:
Looking at definitions of `fraction_lost` and `cumulative number of packets lost`, we should calculate `received` from the beginning, not from the last sender report.  Is this done on purpose? I suspect that we follow libwebrtc or not really? 

Fraction lost is defined as 256 * fraction, where fraction is:

```
This fraction is defined to be the number
      of packets lost divided by the number of packets expected, as
      defined in the next paragraph.
```
      
and the number of packets expected:

```
The number of packets expected is
      defined to be the extended last sequence number received, as
      defined next, less the initial sequence number received.
```

Reply:
The very fist sentence of the definition in RFC 3550 is `The fraction of RTP data packets from source SSRC_n lost since the previous SR or RR packet was sent`. The `Appendix A.3` also explains how to calculate loss from the last reporting interval, not from the beginning. This 100% should be loss from the last reporting interval, also that's how every one implements it. 

When it comes to `cumulative number of packets lost`, we indeed use a number of packets lost since the very beginning and that's how it's implemented here.